Cuphead/Cuphead.asl

```diff
@@ -300,17 +343,37 @@ split
 
 reset
 {
-	if (current.InILMode && (current.Loading && current.Time == 0f || current.InOverworld))
+	if (current.InILMode && (current.Loading && current.LSDTime == 0f || current.InOverworld))
 	{
-		vars.Log("Resetting due to reset {} | Time: " + current.Time + " | Loading: " + current.Loading + " | InOverworld: " + current.InOverworld);
+		vars.Log("Resetting due to reset {} | Time: " + current.Time + " | InOverworld: " + current.InOverworld + " | InKingDice: " + current.InKingDice + " | Loading: " + current.Loading);
+		current.IsKDLevelEnding = false;
 		return true;
 	}
 }
 
 gameTime
 {
 	if (current.InILMode)
-		return TimeSpan.FromSeconds(current.Time);
+	{
+		if (!current.InKingDice)
+			return TimeSpan.FromSeconds(current.Time);
+		
+		// King dice is a series of levels whose time at the end is a sum of all levels, updated after each level is completed.
+		// If a level is ending we just use that time, otherwise we update the current time with the current level time
+		var time = current.IsKDLevelEnding ? current.LSDTime : current.LSDTime + current.Time;
+		return TimeSpan.FromSeconds(time);
+	}
+	
+	// Sometimes the time is 0.01 during the first loading screen, it should always be 0, so we force that here
+	// https://discord.com/channels/144133978759233536/144134231201808385/727009875569279048
+	if(timer.CurrentTime.GameTime.HasValue
+	    && current.Loading
+		&& timer.CurrentTime.GameTime.Value > TimeSpan.Zero
+	    && timer.CurrentTime.RealTime.Value < TimeSpan.FromMilliseconds(50))
+	{
+		vars.Log("Overwriting game time to 0 | RealTime: " + timer.CurrentTime.RealTime.Value + " | GameTime: " + timer.CurrentTime.GameTime.Value);
+		return TimeSpan.Zero;
+	}
```

Comment:
In fact, please completely replace this with `timer.SetGameTime(TimeSpan.Zero);` in `onStart {}`.

Reply:
when does that fire in the event loop? w/rt isLoading - will it catch the discrepancy?

Reply:
> w/rt

Pardon?

Reply:
with respect to

Reply:
I'm not sure what the question is here. the `timer.IsGameTimePaused = true;` line simply isn't executed quickly enough to catch the timer starting occasionally. To work around that, simply set the time to 0.

```cs
onStart
{
	timer.IsGameTimePaused = true;
	timer.SetGameTime(TimeSpan.Zero);
}
```

Reply:
I didn't know if onStart would run before or after isLoading and therefore if it would set it to 0 too early, but setting `timer.IsGameTimePaused` explicitly to true seems to work around this entirely, so seems good